src/iso9660.rs

```diff
@@ -439,38 +440,41 @@ fn get_next_directory_record(buf: &mut Bytes, length: u32) -> Result<Option<Dire
     let length = eat(buf, 4).get_u32_le();
     let flags = eat(buf, 25 - 14).get_u8();
     let name_length = eat(buf, 32 - 26).get_u8() as usize;
-    let name = parse_iso9660_path(buf, name_length).context("parsing record name")?;
+    let name = if is_root && name_length == 1 && buf[0] == 0 {
+        // as a special case, allow "." when reading the root directory
+        // record from the primary volume descriptor
+        eat(buf, name_length);
+        Some(".".into())
+    } else if name_length == 1 && (buf[0] == 0 || buf[0] == 1) {
+        // "." or ".."; ignore
+        eat(buf, name_length);
+        None
+    } else {
+        Some(
+            parse_iso9660_string(buf, name_length, IsoString::File)
+                .context("parsing record name")?,
+        )
+    };
 
     // advance to next record
     eat(buf, len - (33 + name_length));
 
-    if flags & 2 > 0 {
-        Ok(Some(DirectoryRecord::Directory(Directory {
-            name,
-            address,
-            length,
-        })))
-    } else {
-        Ok(Some(DirectoryRecord::File(File {
-            name,
-            address,
-            length,
-        })))
-    }
-}
-
-/// Reads a directory record path. This is similar to a regular ISO9660 string, but supports '\0'
-/// to mean current directory, and '\1' for the parent directory.
-fn parse_iso9660_path(buf: &mut Bytes, len: usize) -> Result<String> {
-    if len == 1 && (buf[0] == 0 || buf[0] == 1) {
-        let c = buf.get_u8();
-        if c == 0 {
-            Ok(".".into())
+    if let Some(name) = name {
+        if flags & 2 > 0 {
+            Ok(Some(DirectoryRecord::Directory(Directory {
+                name,
+                address,
+                length,
+            })))
         } else {
-            Ok("..".into())
+            Ok(Some(DirectoryRecord::File(File {
+                name,
+                address,
+                length,
+            })))
         }
     } else {
-        parse_iso9660_string(buf, len, IsoString::File)
+        get_next_directory_record(buf, length, is_root)
```

Comment:
This recursive call is kind of awkward. I know it's subject to TCO, but... I had the same feeling about the jump to the next 2k alignment at the beginning of the function where we hackily just advance the cursor and pretend to start from scratch. Thinking it'd be cleaner at this point to just move to a loop where the start of each iteration always assumes we're at the start of a directory record.

Reply:
Yup, good call.